nncf/pruning/filter_pruning/algo.py

```diff
@@ -280,62 +282,82 @@ def _find_layerwise_pruning_rate(self, target_flops_pruning_rate):
             return right
-        raise RuntimeError("Can't prune model to asked flops pruning rate = {}".format(target_flops_pruning_rate))
+        raise RuntimeError("Can't prune the model to get the required "
+                           "pruning rate in flops = {}".format(target_flops_pruning_rate))
 
-    def set_pruning_rate(self, pruning_rate, run_batchnorm_adaptation=False):
+    def set_pruning_rate(self, pruning_rate: Union[float, Dict],
+                         run_batchnorm_adaptation: bool = False, groupwise_pruning_rates: bool = False):
         # Pruning rate from scheduler can be flops pruning rate or percentage of params that should be pruned
         self.pruning_rate = pruning_rate
+        group_ids = [group.id for group in self.pruned_module_groups_info.get_all_clusters()]
         if not self.frozen:
-            nncf_logger.info("Computing filter importances and masks...")
-            if self.all_weights:
-                if self.prune_flops:
-                    self._set_binary_masks_for_all_pruned_modules_by_flops_target(pruning_rate)
+            nncf_logger.info("Computing filter importance scores and binary masks...")
+            with torch.no_grad():
+                if self.all_weights:
+                    # Non-uniform (global) importance-score-based pruning according
+                    # to the global pruning rate
+                    if self.prune_flops:
+                        self._set_binary_masks_for_pruned_modules_globally_by_flops_target(pruning_rate)
+                    else:
+                        self._set_binary_masks_for_pruned_modules_globally(pruning_rate)
+                elif groupwise_pruning_rates:
+                    if set(pruning_rate.keys()) != set(group_ids):
+                        raise RuntimeError('Groupwise pruning rate dict keys do not correspond to '
+                                           'layer group ids')
+                    self._set_binary_masks_for_pruned_modules_groupwise(pruning_rate)
                 else:
-                    self._set_binary_masks_for_all_pruned_modules(pruning_rate)
-            else:
-                layerwise_pruning_rate = pruning_rate
-                if self.prune_flops:
-                    # Looking for layerwise pruning rate needed for asked flops pruning rate
-                    layerwise_pruning_rate = self._find_layerwise_pruning_rate(pruning_rate)
-                self._set_binary_masks_for_filters(layerwise_pruning_rate)
+                    # Pruning uniformly with the same pruning rate across layers
+                    layerwise_pruning_rate = pruning_rate
+                    if self.prune_flops:
+                        # Looking for layerwise pruning rate needed for the required flops pruning rate
+                        layerwise_pruning_rate = self._find_uniform_pruning_rate_for_target_flops(pruning_rate)
+                    layerwise_pruning_rate_dict = {group_id: layerwise_pruning_rate for group_id in group_ids}
```

Comment:
May be batter change `_set_binary_masks_for_pruned_modules_groupwise` to work with `pruning_rate` as float and Dict.
```
groupwise_pruning_rates = isinstance(pruning_rate, Dict)
...
pruning_rate[group.id] if groupwise_pruning_rates else pruning_rate  
```

Reply:
Agree, I also prefer one common function to work with both `pruning_rate` formats to reduce code duplication.